sources/api/netdog/src/cli/write_resolv_conf.rs

```diff
@@ -1,28 +1,177 @@
 use crate::networkd_status::NetworkDInterfaceStatus;
-use argh::FromArgs;
-use snafu::ResultExt;
+#[cfg(net_backend = "systemd-networkd")]
+use snafu::{ensure, OptionExt};
+#[cfg(net_backend = "systemd-networkd")]
+static SYSTEMCTL: &str = "/usr/bin/systemctl";
+#[cfg(net_backend = "systemd-networkd")]
+use crate::cli::fetch_net_config;
+#[cfg(net_backend = "systemd-networkd")]
+use crate::networkd::config::NETWORKD_CONFIG_DIR;
+#[cfg(net_backend = "systemd-networkd")]
+use std::{fs, path::Path};
+#[cfg(net_backend = "systemd-networkd")]
+use systemd_derive::{SystemdUnit, SystemdUnitSection};
 
 #[derive(FromArgs, PartialEq, Debug)]
 #[argh(subcommand, name = "write-resolv-conf")]
 /// Writes /etc/resolv.conf, using DNS API settings if they exist
 pub(crate) struct WriteResolvConfArgs {}
 
+/// A struct representing an interface drop-in that overrides DNS-via-DHCP settings
+// TODO: if we end up creating more interface drop-ins that use network settings, it's probably
+// worth making the notion of a drop-in a first class citizen in the networkd config module.
+#[cfg(net_backend = "systemd-networkd")]
+#[derive(Debug, SystemdUnit)]
+struct InterfaceDNSDropIn {
+    dhcp4: Option<Dhcp4Section>,
+    dhcp6: Option<Dhcp6Section>,
+}
+
 #[cfg(net_backend = "systemd-networkd")]
-fn get_dns_settings(primary_interface: String) -> Result<DnsSettings> {
-    let primary_link_status = NetworkDInterfaceStatus::new(primary_interface)
-        .context(error::NetworkDInterfaceStatusSnafu)?;
-    let dns_settings = DnsSettings::from_config_or_status(&primary_link_status)
-        .context(error::GetDnsSettingsSnafu)?;
-    Ok(dns_settings)
+#[derive(Debug, SystemdUnitSection)]
+#[systemd(section = "DHCPv4")]
+struct Dhcp4Section {
+    #[systemd(entry = "UseDNS")]
+    use_dns: Option<bool>,
+    #[systemd(entry = "UseDomains")]
+    use_domains: Option<bool>,
 }
 
+#[cfg(net_backend = "systemd-networkd")]
+#[derive(Debug, SystemdUnitSection)]
+#[systemd(section = "DHCPv6")]
+struct Dhcp6Section {
+    #[systemd(entry = "UseDNS")]
+    use_dns: Option<bool>,
+    #[systemd(entry = "UseDomains")]
+    use_domains: Option<bool>,
+}
+
+#[cfg(net_backend = "systemd-networkd")]
+impl InterfaceDNSDropIn {
+    /// Using API DNS settings, create an appropriate drop-in for a network interface.
+    fn from_dns_settings(settings: &DnsSettings) -> Option<Self> {
+        // If there aren't any settings, don't write a drop-in
+        if !settings.has_name_servers() && !settings.has_search_domains() {
+            return None;
+        }
+
+        // If there are name server or search domain API settings, turn off the option to get these
+        // from the DHCP server.
+        let use_dns = match settings.has_name_servers() {
+            true => Some(false),
+            false => Some(true),
+        };
+        let use_domains = match settings.has_search_domains() {
+            true => Some(false),
+            false => Some(true),
+        };
```

Comment:
The `match` on a boolean is also a bit strange.

```suggestion
        let should_use_domains_from_dhcp = Some(!settings.has_search_domains());
        let should_use_dns_from_dhcp = Some(!settings.has_name_servers());
```
 
```